EpsilonAuxBMS/Src/SetContactorsTask.c

```diff
@@ -0,0 +1,284 @@
+                {
+                    state = prevState;
+                }
+
+                break;
+
+            default:
+                state = FIRST_CHECK;
+        }
+    }
+}
+
+int updateContactorState(ContactorState state, Contactor contactor)
+{
+    if (osMutexWait(auxStatus.auxStatusMutex, 0) != osOK)
+    {
+        return 0;
+    }
+
+    if (contactor == COMMON)
+    {
+        auxStatus.commonContactorState = state;
+    }
+    else if (contactor == CHARGE)
+    {
+        auxStatus.chargeContactorState = state;
+    }
+    else if (contactor == DISCHARGE)
+    {
+        auxStatus.dischargeContactorState = state;
+    }
+
+    osMutexRelease(auxStatus.auxStatusMutex);
+    return 1;
+}
+
+int checkIfContactorSet(uint16_t pin, GPIO_TypeDef* port, int current_multiplier)
+{
+    float pwrVoltage = 0.0;
+    float pwrCurrent = 0.0;
+    uint8_t sense_pin;
+    // Get 12bit current analog input
+    uint32_t current_sense = readCurrentThroughContactors();
+
+    if (current_sense == 0xFFFFFFFF)
+    {
+        return -1;
+    }
+
+    pwrVoltage = 3.3 * current_sense / 0xFFF / GAIN; // change ADC value into the voltage read
+    pwrCurrent = pwrVoltage / CURRENT_SENSE_RESISTOR; // convert voltage to current
+
+    if (pin == 0xFF && port == NULL)
+    {
+        sense_pin = 1;
+    }
+    else
+    {
+        sense_pin = !HAL_GPIO_ReadPin(port, pin);
+    }
+
+    // If current is high for some reason or sense pin is high something is wrong
+    if (pwrCurrent > CURRENT_LOWER_THRESHOLD * current_multiplier || !sense_pin)
+    {
+        return -1;
+    }
+
+    return 0;
+}
+
+/*
+  This function takes multiple reads because the current amplifier has a bit of a setup time
+  to make sure that erroneous values aren't read, multiple reads are taken till we have a
+  relatively steady value. Dan suggested that 3 good reads in a row is good.
+*/
+uint32_t readCurrentThroughContactors(void)
+{
+    // Enable current amplifier. Allow time to settle and take multiple valid concurrent readings
+    // to make sure it isn't changing
+    HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_SET);
+    uint32_t sense = 0;
+    uint32_t prev_sense = 0;
+
+    for (int counter = 0; counter < 3; counter++)
+    {
+        osDelay(SENSE_SETUP_TIME);
+
+        if (HAL_ADC_PollForConversion(&hadc1, ADC_POLL_TIMEOUT) == HAL_OK)
+        {
+            sense = HAL_ADC_GetValue(&hadc1);
+            int diff = sense - prev_sense;
+
+            if (diff < 0)
+            {
+                diff *= -1;
+            }
+
+            if (diff > 50)
+            {
+                prev_sense = sense;
+                counter = -1; // Will be incremented to 0 at the end of the loop
+            }
+        }
+        else
+        {
+            sense = 0xFFFFFFFF;
+            break;
+        }
+    }
+
+    HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_RESET);
```

Comment:
Because of the osDelay for hte sense setup time, the ADC is going to be held for alot longer than it needs to. This pin set and reset should be done once for everytime the value, not once for the whole function.

Reply:
I can't reset that pin because if I do, it defeats the purpose of the function. By doing the setting and resetting, I'm turning the current amplifier on and off. So we're trying to get good reads after turning it on. Then once we get the consistency, it can go off.